### DISCUSSION TITLE
When a session is aborted by server end, how can re-connect to server end with this origin session?

### DISCUSSION THREAD

**Discussion by flyly0755** [sqlalchemy/sqlalchemy#10875] [Usage Questions] [answered]
### Describe the use case

Now I use clickhouse database to store data, but with spontaneous disconnects issue from server side detailed in the page showed below:
https://github.com/ClickHouse/ClickHouse/issues/52571
what i want is before crud i take a try session.execute("select 1")
if raise error:
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
then i re-connect this session to make this session available again.

### Databases / Backends / Drivers targeted

database: clickhouse
python package: 
SQLAlchemy==1.4.49
SQLAlchemy-Utils==0.41.1
clickhouse-sqlalchemy==0.2.3

### Example Use

i try to print session.is_active which is still True when issue arises, i wrongly think which maybe False.
so i try with try except method to catch this issue, 
```python
try:
    session.execute("select 1")
except Exception as e:
   ---how to reconnect this session to make it available again?
```
a method will be:
```python
uri = "clickhouse://{}:{}@{}:{}/{}".format(user, pwd, ip, port, dbname)  
engine = create_engine(uri, echo=False, poolclass=NullPool)
session = sessionmaker(engine)()
```
re-generate a new session to replace the  out-of-date session.

but in some situation, session is a function parameter, so can't get uri infomation.
So i think is there some simpler method to achieve it?
like session.reactive() or something else?

**Comment by zzzeek** [comment, marked as answer]
the `Session` is not where the connectivity is, it's in the `Connection` that is held onto by the `Session`.     if you have a connectivity issue, that by definition cancels the transaction which the `Session` is participating within, so at the very least you have to call `Session.rollback()` when an exception occurs that cancels the transaction.  once you do that, the `Session` will be able to begin a new transaction on a new connection.

There is no way in the `Session` to reconnect without the transaction being restarted.    You can use [driver level autocommit](https://docs.sqlalchemy.org/en/20/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit) on the connection so that there is not any actual ROLLBACK emitted to the database, however, when using the higher level `Session` abstraction, there's always a `SessionTransaction` concept that will expire all the state in your ORM objects that are associated with it.

so there's some things to consider here but overall if you wanted to "restart" the `Session` inside your method, you would just do a rollback() and then the `Session` will continue working:

```
try:
    session.execute("select 1")
except Exception as e:
    session.rollback()
    
    # session is ready to use again
```

as far as there being a rollback in the middle of the method, if you lost connectivity, your transaction was lost in any case.

**Comment by flyly0755** [comment]
session.rollback() is useful to reset the session， thx:)
